mypyc/genops.py

```diff
@@ -584,27 +540,25 @@ def gen_func_def(self, fdef: FuncDef, sig: FuncSignature,
         nested function, then we generate an environment class so that inner nested functions can
         access the environment of the given FuncDef.
         """
-        self.fn_info = FuncInfo(fdef)
+        self.fn_info = FuncInfo(fdef, self.gen_func_ns())
         self.fn_infos.append(self.fn_info)
 
         # If there is more than one environment in the environment stack, then we are visiting a
         # non-global function (the top-most environment is for the module top level).
         self.fn_info.is_nested = len(self.environments) > 2
         self.fn_info.contains_nested = self.contains_func_def(fdef)
 
-        namespace = self.gen_func_ns()
-
         self.enter(fdef.name())
 
         if self.fn_info.is_nested:
-            self.setup_callable_class(fdef, namespace)
+            self.setup_callable_class()
         if self.fn_info.contains_nested:
-            self.setup_env_class(fdef, namespace)
+            self.setup_env_class()
 
-        self.load_env_registers(fdef)
+        self.load_env_registers()
```

Comment:
I'm not particularly convinced that the refactor to make more state implicit in the `FuncInfo` stack makes the code easier to read and modify and not harder, though I'm also not particularly convinced it makes it harder. `IRBuilder` is already super heavy on the implicit state anyways.

As a procedural issue, I think that the refactor probably ought to have been submitted as a separate PR. The free variable finding changes seem like a nice elegant diff that is cluttered up by a noisy refactor.